File: svws-webclient/core/src/core/utils/uvblockung/UvAlgorithmusDynDatenRegel27.ts
import { JavaObject } from '../../../java/lang/JavaObject';
import { UvAlgorithmusDynDatenKlasse } from '../../../core/utils/uvblockung/UvAlgorithmusDynDatenKlasse';
import { UvAlgorithmusDynDatenLehrkraft } from '../../../core/utils/uvblockung/UvAlgorithmusDynDatenLehrkraft';
import type { UvAlgorithmusDynDatenRegel } from '../../../core/utils/uvblockung/UvAlgorithmusDynDatenRegel';
import { cast_de_svws_nrw_core_utils_uvblockung_UvAlgorithmusDynDatenRegel } from '../../../core/utils/uvblockung/UvAlgorithmusDynDatenRegel';
import { Class } from '../../../java/lang/Class';

export class UvAlgorithmusDynDatenRegel27 extends JavaObject implements UvAlgorithmusDynDatenRegel {

	private readonly malus: Array<number>;

	private readonly klasse: UvAlgorithmusDynDatenKlasse;

	private readonly lehrkraft1: UvAlgorithmusDynDatenLehrkraft;

	private readonly lehrkraft2: UvAlgorithmusDynDatenLehrkraft;

	private readonly prioritaet: number;


	/**
	 * Erzeugt eine neue Instanz der Regel und passt den Anfangs-Malus an.
	 *
	 * @param malus        Das globale Malus-Array.
	 * @param klasse       Die {@link UvAlgorithmusDynDatenKlasse}.
	 * @param lehrkraft1   Die 1. {@link UvAlgorithmusDynDatenLehrkraft}.
	 * @param lehrkraft2   Die 2. {@link UvAlgorithmusDynDatenLehrkraft}.
	 * @param prioritaet   Die Priorität der Regel.
	 */
	public constructor(malus: Array<number>, klasse: UvAlgorithmusDynDatenKlasse, lehrkraft1: UvAlgorithmusDynDatenLehrkraft, lehrkraft2: UvAlgorithmusDynDatenLehrkraft, prioritaet: number) {
		super();
		this.malus = malus;
		this.klasse = klasse;
		this.lehrkraft1 = lehrkraft1;
		this.lehrkraft2 = lehrkraft2;
		this.prioritaet = prioritaet;
		this.changeMalus(1);
	}

	public changeMalus(faktor: number): void {
		const klasseID: number = this.klasse.interneID;
		if ((this.prioritaet >= 0) && (this.lehrkraft1.istStundenProKlasse[klasseID]) > 0 && (this.lehrkraft2.istStundenProKlasse[klasseID] > 0)) {
			this.malus[this.prioritaet] += faktor;
		}
	}

	transpilerCanonicalName(): string {
		return 'de.svws_nrw.core.utils.uvblockung.UvAlgorithmusDynDatenRegel27';
	}

	isTranspiledInstanceOf(name: string): boolean {
		return ['de.svws_nrw.core.utils.uvblockung.UvAlgorithmusDynDatenRegel', 'de.svws_nrw.core.utils.uvblockung.UvAlgorithmusDynDatenRegel27'].includes(name);
	}

	public static readonly class = new Class<UvAlgorithmusDynDatenRegel27>('de.svws_nrw.core.utils.uvblockung.UvAlgorithmusDynDatenRegel27');

}

export function cast_de_svws_nrw_core_utils_uvblockung_UvAlgorithmusDynDatenRegel27(obj: unknown): UvAlgorithmusDynDatenRegel27 {
	return obj as UvAlgorithmusDynDatenRegel27;
}

File: svws-core/src/main/java/de/svws_nrw/core/utils/uvblockung/UvAlgorithmusDynDatenRegel27.java
package de.svws_nrw.core.utils.uvblockung;

import de.svws_nrw.core.data.uv.regel.UvBlockungRegelTyp;
import jakarta.validation.constraints.NotNull;

/**
 * Implementiert die Regel {@link UvBlockungRegelTyp#LEHRKRAEFTE_A_UND_B_NICHT_IN_DER_SELBEN_KLASSE}.
 *
 * @author Benjamin A. Bartsch
 */
public final class UvAlgorithmusDynDatenRegel27 implements UvAlgorithmusDynDatenRegel {

	private final @NotNull int @NotNull [] malus;
	private final @NotNull UvAlgorithmusDynDatenKlasse klasse;
	private final @NotNull UvAlgorithmusDynDatenLehrkraft lehrkraft1;
	private final @NotNull UvAlgorithmusDynDatenLehrkraft lehrkraft2;
	private final int prioritaet;


	/**
	 * Erzeugt eine neue Instanz der Regel und passt den Anfangs-Malus an.
	 *
	 * @param malus        Das globale Malus-Array.
	 * @param klasse       Die {@link UvAlgorithmusDynDatenKlasse}.
	 * @param lehrkraft1   Die 1. {@link UvAlgorithmusDynDatenLehrkraft}.
	 * @param lehrkraft2   Die 2. {@link UvAlgorithmusDynDatenLehrkraft}.
	 * @param prioritaet   Die Priorität der Regel.
	 */
	public UvAlgorithmusDynDatenRegel27(
			final @NotNull int @NotNull [] malus,
			final @NotNull UvAlgorithmusDynDatenKlasse klasse,
			final @NotNull UvAlgorithmusDynDatenLehrkraft lehrkraft1,
			final @NotNull UvAlgorithmusDynDatenLehrkraft lehrkraft2,
			final int prioritaet) {

		this.malus = malus;
		this.klasse = klasse;
		this.lehrkraft1 = lehrkraft1;
		this.lehrkraft2 = lehrkraft2;
		this.prioritaet = prioritaet;

		this.changeMalus(1);
	}


	@Override
	public void changeMalus(final int faktor) {
		final int klasseID = klasse.interneID;

		// Beide Lehrkräfte müssen Stunden in dieser Klasse haben
		if ((prioritaet >= 0) && (lehrkraft1.istStundenProKlasse[klasseID]) > 0 && (lehrkraft2.istStundenProKlasse[klasseID] > 0)) {
			malus[prioritaet] += faktor;
		}
	}

}
